File: frontend/src/types/card.ts
export type SourceType = "scan" | "pdf" | "url";

export interface BusinessCard {
  id: number;
  person_name: string | null;
  company_name: string | null;
  designation: string | null;
  phone: string | null;
  email: string | null;
  website: string | null;
  instagram: string | null;
  facebook: string | null;
  linkedin: string | null;
  location: string | null;
  gst_number: string | null;
  company_logo: string | null;
  card_image_url: string | null;
  qr_raw: string | null;
  other_details: string | null;
  source_type: SourceType;
  original_file_url?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ExtractedCard {
  owner_name: string;

  job_title?: string | null;

  company_name?: string | null;

  address?: string | null;

  email?: string | null;

  phone?: string | null;

  gst_number?: string | null;

  company_logo?: string | null;

  website_url?: string | null;

  instagram_url?: string | null;

  facebook_url?: string | null;

  linkedin_url?: string | null;

  front_image_url?: string | null;

  back_image_url?: string | null;

  qr_raw?: string | null;

  other_details?: string | null;

  source_type?: string | null;

  original_file_url?: string | null;
}

export interface ApiResponse<T> {
  success: boolean;
  message?: string;
  data?: T;
  card?: T;
}

File: backend/models/card.py
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


# =====================================================
# QR DETAIL
# =====================================================

class QrDetail(BaseModel):
    """
    Structured information about one detected QR code.
    """

    raw: str

    type: str

    label: str

    url: Optional[str] = None


# =====================================================
# CREATE BUSINESS CARD
# =====================================================

class CardCreate(BaseModel):
    owner_name: str

    company_name: Optional[str] = None
    designation: Optional[str] = None

    address: Optional[str] = None

    email: Optional[str] = None
    phone: Optional[str] = None

    gst_number: Optional[str] = None
    company_logo: Optional[str] = None

    instagram_url: Optional[str] = None
    website_url: Optional[str] = None
    facebook_url: Optional[str] = None
    linkedin_url: Optional[str] = None

    other_details: Optional[str] = None

    # =====================================================
    # QR CODE DATA
    # =====================================================

    # -----------------------------------------------------
    # Backward-compatible single QR value
    # -----------------------------------------------------
    #
    # Older records / older scanner code may still use this.
    #
    qr_raw: Optional[str] = None

    # -----------------------------------------------------
    # ALL DETECTED QR CODES
    # -----------------------------------------------------
    #
    # Example:
    #
    # qr_codes = [
    #     "https://www.revibeperfume.com/",
    #     "https://www.instagram.com/revibeperfume/"
    # ]
    #
    qr_codes: Optional[list[str]] = Field(
        default=None
    )

    # -----------------------------------------------------
    # STRUCTURED QR DETAILS
    # -----------------------------------------------------
    #
    # Example:
    #
    # [
    #     {
    #         "raw": "https://www.revibeperfume.com/",
    #         "type": "website",
    #         "label": "Website",
    #         "url": "https://www.revibeperfume.com/"
    #     },
    #     {
    #         "raw": "https://www.instagram.com/revibeperfume/",
    #         "type": "instagram",
    #         "label": "Instagram",
    #         "url": "https://www.instagram.com/revibeperfume/"
    #     }
    # ]
    #
    qr_details: Optional[list[QrDetail]] = Field(
        default=None
    )

    # =====================================================
    # PYDANTIC CONFIG
    # =====================================================

    model_config = ConfigDict(
        extra="ignore"
    )


# =====================================================
# URL REQUEST
# =====================================================

class UrlRequest(BaseModel):
    url: str